src/network/networkd-network.c

```diff
@@ -36,6 +36,70 @@
 #include "string-util.h"
 #include "util.h"
 
+static void network_config_hash_func(const void *p, struct siphash *state) {
+        const NetworkConfigSection *c = p;
+
+        siphash24_compress(c->filename, c->filename_size, state);
+        siphash24_compress(&c->filename_size, sizeof(c->filename_size), state);
+        siphash24_compress(&c->line, sizeof(c->line), state);
+}
+
+static int network_config_compare_func(const void *a, const void *b) {
+        const NetworkConfigSection *x = a, *y = b;
+        int r;
+
+        r = memcmp(x->filename, y->filename, MIN(x->filename_size, y->filename_size));
+        if (r != 0)
+                return r;
+
+        if (x->filename_size < y->filename_size)
+                return -1;
+
+        if (x->filename_size > y->filename_size)
+                return 1;
+
+        if (x->line < y->line)
+                return -1;
+
+        if (x->line > y->line)
+                return 1;
+
+        return 0;
+}
+
+const struct hash_ops network_config_hash_ops = {
+        .hash = network_config_hash_func,
+        .compare = network_config_compare_func
+};
+
+int network_config_section_new(const char *filename, unsigned line, NetworkConfigSection **s) {
+        _cleanup_network_config_section_free_  NetworkConfigSection *n = NULL;
+
+        n = new0(NetworkConfigSection, 1);
+        if (!n)
+                return -ENOMEM;
+
+        n->filename = strdup(filename);
+        if (!n->filename)
+                return -ENOMEM;
+
+        n->filename_size = strlen(filename);
```

Comment:
strlen() is actually relatively cheap. memory allocations are not. i'd really focus on avoiding memory allocations, and reagrding the string sizes I'd try to make things simple.